### PR TITLE
[GraphBolt] Lock-free `CachePolicy::ReadWriteCompleted`.

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
The ReadingCompleted and WritingCompleted operations are made lock-free so we don't have to take the lock before calling them anymore.

Compiled with thread-sanitizer and checked for data races. Everything seems to simply work.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/src/cache_policy.h [hunk 1]
```diff
@@ -79,35 +83,58 @@ struct CacheKey {
...
-  int64_t freq_ : 3;
+  int8_t freq_;
   // Negative values indicate writing while positive values indicate reading.
-  int64_t reference_count_ : 13;
-  int64_t key_ : 48;
+  // Access only through an std::atomic_ref instance atomically.
+  int8_t reference_count_;
+  // Keys are restricted to be 48-bit unsigned integers.
+  uint16_t key_upper_;
+  uint32_t key_lower_;
```
Comment (frozenbugs) [inline graphbolt/src/cache_policy.h:137]: Can you explain what's key upper and key lower?

Comment (mfbalin) [reply graphbolt/src/cache_policy.h:137]: Keyupper is upper 16 bits of key, keylower is lower 32 bits. Isn't it clear in the code?

Comment (mfbalin) [reply graphbolt/src/cache_policy.h:137]: Maybe I can rename the variables to `key_upper_16_bits_` and `key_lower_32_bits_` instead.

Comment (mfbalin) [reply graphbolt/src/cache_policy.h:137]: #7659

**Pull request merged**